Support for Modular RCS

al2me6 opened the pull request:
Notably, `RCSDATA` has two additional keys `thrustTransformPositionOffset` (Vector3) and `thrustTransformScaleOffset` (float). These are necessary for proper (Waterfall) plume support.

Consider a modular RCS part with two models A and B, with B having a nozzle twice as big as A at the same scale. Thus, B needs to be at half scale to match A. On the other hand, the plume isn't aware of this scaling. Thus, since B is at half-scale, so is its thrust transform, and the plume, which is the right size for a, comes out as half the correct size on B. In this case, a `thrustTransformScaleOffset` of `2` applied to B will undo the outer rescaling and produce the right plume size.

`thrustTransformPositionOffset` is useful for fixing models with incorrectly positioned thrust transforms.

pap1723 committed:
Begin the process of adding Modular RCS

al2me6 committed:
Remove ROEngines project

al2me6 committed:
Remove stat calculations from model definition

al2me6 committed:
Remove more logging

al2me6 committed:
Allow RCS thrust transforms to be moved & scaled
This is necessary for proper plume scaling.

DRVeyl commented on Source/ROLib/ModelDef/ROLModelModule.cs:
```diff
@@ -389,7 +389,14 @@ public void setupModel(bool doNotRescaleX = false)
+                return;
+            }
+            definition.rcsModuleData.RenameTransforms(root, destinationName);
+        }
```
Why this change?  These are not Unity-managed classes and null-coalescing is valid here.

DRVeyl commented on Source/ROLib/ModelDef/ROLModelModule.cs:
```diff
@@ -427,25 +434,17 @@ public void UpdateEngineModuleThrust(ModuleEngines engine, float thrustScalePowe
             }
-            rcs.thrusterPower = power;
-            rcs.moduleIsEnabled = power > 0;
             rcs.OnStart(PartModule.StartState.Flying);
         }
```
Can you guarantee this is only ever called after the normal rcs OnStart?  And not at potentially inappropriate times such as before it has Loaded?

DRVeyl replied:
Actually, I believe this doesn't have any place in the model manipulation.  It may have been relevant for setting the power values.  But that's now properly handled elsewhere, and so should whatever fixup is necessary to the running state be.

al2me6 replied:
This I believe was inherited from whatever was in SSTU. But yes; I agree that it's quite fishy and should be moved. In fact there's code in ROE working _around_ this call.

DRVeyl replied:
I'll remove the entire method here.  Setting those enabled fields should go into the ROE code to update the module using the data in the definition.

DRVeyl commented on Source/ROLib/ModelDef/ROLModelDefinition.cs:
```diff
@@ -1,1096 +1,1115 @@
-﻿using System;
```
Why is this *entire file* diff'd?  Not sure how to review this.

siimav replied:
I imagine inconsistent line endings were normalized

DRVeyl commented:
I guess I can go through the individual commits, though the conflicts in the branch need to be resolved.

DRVeyl committed:
Don't expand RenameRCSThrustTransforms

DRVeyl committed:
Remove UpdateRCSModule entirely

These RCS-specific functions should go into an RCS module.  These are not really model-based manipulations.

DRVeyl committed:
Fix Bizarrely Complex Code

Don't write a [bad] custom wrapper of Contains, and use Contains within it.
What's a few extra foreach among friends?

DRVeyl committed:
Minor code tweaks

DRVeyl committed:
Fix disableTransforms setup

DRVeyl commented:
Ok.  Rebased on master, cleaned up some things, removed the commits of the dll [those can be done on master branch when it is time for release or as others prefer, but avoid doing them in feature branches and committing at least because of the conflicts it creates.]

@al2me6 Does this look sane?

pap1723 committed:
Add modelScale value to set the correct starting size of RCS Parts

pap1723 committed:
Removed duplicated line from Copypasta

pap1723 committed:
Remove GameEvent that was called everytime the craft was edited

pap1723 committed:
Remove modelScale and properly use scaleOffset instead

NathanKell committed:
Update dll for Stations too
